rust/src/daemon.rs

```diff
@@ -246,6 +254,104 @@ pub(crate) fn deployment_layeredmeta_load(
+
+    let (ostree_remote, _) = ostree::parse_refspec(refspec)?;
+    let mut gpg_verify = false;
+    if let Some(ref ostree_remote) = ostree_remote {
+        match repo.remote_get_gpg_verify(ostree_remote.as_str()) {
+            Ok(b) => gpg_verify = b,
+            Err(err) => {
+                if let Some(ioerr) = err.kind::<gio::IOErrorEnum>() {
+                    // If the remote doesn't exist, let's note that so that status can render it
+                    // specially.
+                    if ioerr == gio::IOErrorEnum::NotFound {
+                        let msg = format!("{}", err);
+                        dict.insert("remote-error", &msg);
+                        return Ok(());
+                    }
+                }
+                return Err(err.into());
+            }
+        };
+    }
+
+    dict.insert("gpg-enabled", &gpg_verify);
+    if !gpg_verify {
+        return Ok(()); // note early return; no need to verify signatures!
+    }
+
+    // there must a remote if GPG checking is enabled
+    let ostree_remote = ostree_remote.unwrap();
+
+    match get_cached_signatures_variant(repo, ostree_remote.as_str(), base_checksum) {
+        Ok(sigs) => dict.insert("signatures", &sigs),
+        Err(err) => {
+            // Somehow, we have a deployment which has gpg-verify=true, but we couldn't verify its
+            // signature. Let's not just bomb out here. We need to return this in the variant so
+            // that `status` can show the appropriate msg.
+            tracing::debug!("failed to get cached signature variant: {}", err);
+        }
+    };
+
+    Ok(())
+}
+
+fn get_cached_signatures_variant(
+    repo: &ostree::Repo,
+    remote: &str,
+    checksum: &str,
+) -> Result<glib::Variant> {
+    std::fs::create_dir_all(RPM_OSTREED_COMMIT_VERIFICATION_CACHE)?;
+    let dir = cap_std::fs::Dir::from_std_file(std::fs::File::open(
+        RPM_OSTREED_COMMIT_VERIFICATION_CACHE,
+    )?);
+
+    let remote_dir = Path::new(remote);
+    let cached_relpath = remote_dir.join(checksum);
+
+    if let Some(mut f) = dir.open_optional(&cached_relpath)? {
+        let m = f.metadata()?;
+        if m.file_type().is_file() && m.mode() == 0o600 && m.uid() == 0 && m.gid() == 0 {
+            tracing::debug!("signature variant cache hit for ({}, {})", remote, checksum);
+            let bytes = {
+                let mut buf: Vec<u8> = Vec::new();
+                f.read_to_end(&mut buf)?;
+                glib::Bytes::from_owned(buf)
+            };
+            let v = glib::Variant::from_bytes::<Vec<glib::Variant>>(&bytes);
+            return Ok(v);
+        }
+    }
+
+    tracing::debug!(
+        "signature variant cache miss for ({}, {})",
+        remote,
+        checksum
+    );
+
+    dir.create_dir_all(remote_dir)?;
+    dir.remove_file_optional(&cached_relpath)?;
+    let verify_result = repo.verify_commit_for_remote(checksum, remote, gio::NONE_CANCELLABLE)?;
+    let mut sigs = Vec::new();
```

Comment:
```suggestion
    let mut sigs = Vec::with_capacity(verify_result.count_all());
```

(Hmm also we should change the verify result API to `impl Iterator`)